# Command Line Basics & Assembly Prep

This week, we mostly touched on basics of shell (bash) scripting, common commands, and for loops.  To have a sense of what we were working with, we also prepped assemblies for four publicly-available *S. aureus* genomes.

---

## *S. aureus* genome prep

To have a better sense of the genomes that we have, we:

- Downloaded reads from NCBI [fasterq-dump](https://rnnh.github.io/bioinfo-notebook/docs/fasterq-dump.html)
- Ran trimomatic [trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic)
- Performed *de novo* assembly with [SPAdes](https://github.com/ablab/spades)

### Download reads

Reads can be downloaded from NCBI efficiently with fasterq-dump (note this is a different tool from fastq-dump, which allows multi-threading).

fasterq-dump is a tool within SRA Tools, which can be installed as a conda environment, after which, the following can be run:

``` console
# set values for variables in generalized code
IDlist=/home/mwoodwo/prj/Berryhill/99.lists/IDlist.txt
outdir=/home/mwoodwo/prj/Berryhill/00.raw

for ID in `cat $IDlist`; do echo starting $ID ...; fasterq-dump $ID -t ${outdir}/fastq -e 10; echo $ID seems complete ...; done

# with comma-delimited ID file SAMPLE,SRR_ACC
for line in `cat $IDlist`; do sample=$(echo $line | cut -d, -f1); ID=$(echo $line | cut -d, -f2); echo sample is $sample    accession is $ID; if [ ! -d "${outdir}/${sample}" ]; then mkdir ${outdir}/${sample}; echo ''; echo "====================================="; echo "created ${outdir}/${sample}"; fi; echo starting $ID ...; fasterq-dump -O ${outdir}/${sample} -e 10 -p $ID; echo $ID seems complete ...; echo ''; done

```

This should produce a a directory in your specified outdir for each sample, with sub-directories for each library of reads.

We can then concatenate the 1/2 reads for each library into one sample read file as follows:

``` console
# loop over IDlist and concatenate _1.fastq and _2.fastq files for each library

for line in `cat $IDlist`; do sample=$(echo $line | cut -d, -f1); ID=$(echo $line | cut -d, -f2); echo sample is $sample    accession is $ID; cat ${sample}/${ID}_1.fastq >> ${sample}/${sample}_1.fastq; cat ${sample}/${ID}_2.fastq >> ${sample}/${sample}_2.fastq; done
```
