# Polished/Standalone bacterial genomic workflows

Week 4 goals are to introduce PATRIC and Bactopia, which are two useful tools for bacterial genomics that have well-documented workflows.

---

## PATRIC

[PATRIC](https://patricbrc.org) is the Pathosystems Resource Integration Center focused on bacterial infectious diseases and based at the University of Chicago.

It's not the most intuitive interface in the world but has workflows that perform assembly, genome annotation, comparative genomic analyses, and construction of phylogenetic trees.

It also is fairly well-documented with tutorials and videos for how to use its features and provides boilerplate text for the methods used that can be included in your manuscript.

- We can run the 'Comprehensive Genome Analysis Service' on your genomes:

| Sample | Accession   |
| ------ | ---------   |
| NRS225 | SRR2910954  |
| NRS225 | SRR2124643  |
| NRS111 | SRR10512470 |
| NRS111 | SRR2910921  |
| NRS111 | SRR2124587  |
| NRS242 | SRR2124656  |
| NRS242 | SRR2910965  |
| NRS112 | SRR10512469 |
| NRS112 | SRR2910922  |
| NRS112 | SRR2124588  |

Other services to check out on PATRIC:
- Similar genome finder service
- Phylogenetic Tree
- Genome alignment

## Bactopia

[Bactopia](https://bactopia.github.io) development was led by Robert Petit, who was at Emory but moved on to Montana. He continues to maintain the tool with regular updates. All software it uses has conda packages, which are handled by [Nextflow](https://www.nextflow.io) pipeline/workflow scripts under the hood.

- Install Bactopia with the [quick start](https://bactopia.github.io/quick-start/).
- 
