# Assemblies & Transfers

Week 3 goals are to evaluate quality of assemblies, predict proteins, annotate proteins, modest pangenomes. Time permitting, we will also review some methods for file transfer to a remote compute node.

---

## Assembly Background

[Touch on some basic points](AssemblyBackground.pptx) about assembly, read mapping, and evaluating their quality.

## Evaluate Assembly Quality

- Install QUAST with conda
- Run QUAST with loop

``` console
quast.py -o ${outdir} -t 8 --glimmer ${indir}/contigs.fasta ${indir}/contigs.fasta ${indir}/contigs.fasta ${indir}/contigs.fasta 
```

## Predict Proteins

- Install Prodigal with conda
- Run Prodigal with loop

``` console
IDlist=/home/mwoodwo/prj/Berryhill/99.lists/IDlist.txt
indir=/home/mwoodwo/prj/Berryhill/02.spades
outdir=/home/mwoodwo/prj/Berryhill/03.prodigal
args='-p single'       # for single genome gene prediction

# for de novo assembled genomes
for ID in `cat $IDlist`; do echo starting $ID ...; scaffold=${indir}/${ID}/scaffolds.fasta; prodigal -a ${outdir}/${ID}.faa -d ${outdir}/${ID}.fna -f gff -i ${scaffold} -o ${outdir}/${ID}.gff $args; echo $ID complete ...; echo ''; done

# for complete assembled genomes (e.g. from MiGS)
for ID in `cat $IDlist`; do echo starting $ID ...; scaffold=${indir}/${ID}.fasta; prodigal -a ${outdir}/${ID}.faa -d ${outdir}/${ID}.fna -f gff -i ${scaffold} -o ${outdir}/${ID}.gff $args; echo $ID complete ...; echo ''; done
```

## Antimicrobial resistance genes analyses (AMRFinder)
[AMRFinderPlus](https://github.com/ncbi/amr) is an NCBI developed/maintained tool that has a robust method of detecting antimicrobial resistance (AMR) genes with blast queries and hidden markov model searches. We have also [written some additional filtering and summarizing scripts](https://github.com/michaelwoodworth/AMRFinder_scripts) that may streamline some analyses with the output of AMRFinder. **We used AMRFinderPlus v3.10.18.**

- Define tool / step variables
```console
indir=${path_to_prodigal_output}
outdir=${path_to_amrfinder_output}
protein=${indir}/${ID}.faa
```

- Run AMRFinder
```console
for ID in `cat $IDlist`; do echo starting $ID ...; protein=${indir}/${ID}.faa; amrfinder -p ${protein} --plus -o ${outdir}/${ID}_amrfinder.tsv; echo $ID complete ...; echo ''; done
```

## Annotation (prokka)
[prokka](https://github.com/tseemann/prokka) is a rapid prokaryotic genome annotation tool. It accepts contigs (produced in earlier assembly step by SPAdes), predicts proteins, and annotates predicted genes. The output can be used directly as input for roary for comparative genomic analyses. **We used Prokka v1.14.6.**

- Define tool / step variables
```console
indir=${path_to_contigs}
outdir=${path_to_prokka_output}
contigs=${indir}/${ID}/contigs.fasta
```

- Run Prokka
```console
for ID in `cat $IDlist`; do echo starting $ID ...; contigs=${indir}/${ID}/contigs.fasta; prokka --outdir ${outdir}/${ID} --prefix $ID $contigs; echo $ID complete ...; echo ''; done
```

## Clustered Gene Analysis (Roary)
[Roary](https://github.com/sanger-pathogens/Roary) is a fairly simple tool for clustered gene comparisons across prokaryote genomes. The simplest way to run Roary is to copy/link gff files of interest from Prokka in a single directory and running Roary from that directory. Unfortunately, it is no longer being supported by the developer. **We used Roary v3.13.0.**

- Define tool / step variables
```console
indir=${path_to_gff_files_from_prokka}
args="-p 8 -o ${ID}_clustered_proteins -e -v"
```

- Run Roary
```console
for ID in `cat $IDlist`; do echo starting $ID ...; roary $args ${indir}/*gff; echo $ID complete ...; echo ''; done
```

## Average Nucleotide Identity (ANI) Tests (FastANI)
[FastANI](https://github.com/ParBLiSS/FastANI) is a fast and simple tool to calculate average nucleotide identity for complete or draft genomes. **We used v1.32.**

It's best to review the documentation on the FastANI github site, but we created a list of all genomes and ran all vs all comparisons as below.

- Define tool / step variables
```console
indir=${path_to_isolate_genome_contigs}
outdir=${path_to_write_output}
genome_list=${path_to_text_file_list_of_genomes}
```

- Run FastANI
```console
for ID in `cat $IDlist`; do echo starting $ID ...; fastANI --ql $genome_list --rl $genome_list -o ${outdir}/output.txt; echo $ID complete ...; echo ''; done
```

## File Transfer

FTP (file transfer protocol) is an unsecured file transfer standard that is occasionally used in public datasets but would avoid this for security reasons.

The easiest way to transfer files with ssh (which encrypts tansfer) is with a desktop client like [Cyberduck](https://cyberduck.io/download/).

Also possible to transfer via the command line, which is necessary in some situations. The most common tools for this are rsync and scp. In addition to transfering files, rsync can check dates & file sizes to evaluate if files need updating. This saves time/bandwidth by not overwriting files that don't need updating. We'll use some rsync examples.

``` console
# Basic syntax
rsync options source destination

# Local to Remote
rsync -avzh ${local_dir} ${username}@170.140.234.60:${remote_dir}/

# Remote to Local
rsync -avzh ${username}@170.140.234.60:${remote_dir} ${local_dir}
```
