# Command Line Basics & Assembly Prep

This week, we mostly touched on basics of shell (bash) scripting, common commands, and for loops.  To have a sense of what we were working with, we also prepped assemblies for four publicly-available *S. aureus* genomes.

---

## *S. aureus* genome prep

To have a better sense of the genomes that we have, we:

- Downloaded reads from NCBI [fasterq-dump](https://rnnh.github.io/bioinfo-notebook/docs/fasterq-dump.html)
- Ran trimomatic [trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic)
- Performed *de novo* assembly with [SPAdes](https://github.com/ablab/spades)

### Download reads

Reads can be downloaded from NCBI efficiently with fasterq-dump (note this is a different tool from fastq-dump, which allows multi-threading).

fasterq-dump is a tool within SRA Tools, which can be installed as a conda environment, after which, the following can be run:

``` console
# set values for variables in generalized code
IDlist=
outidr=

for ID in `cat $IDlist`; do echo starting $ID ...; fasterq-dump $ID -t ${outdir}/fastq -e 10; echo $ID seems complete ...; done

```
